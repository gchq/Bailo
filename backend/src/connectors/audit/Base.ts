import { Request } from 'express'

import { AccessRequestDoc } from '../../models/AccessRequest.js'
import { FileInterface, FileInterfaceDoc } from '../../models/File.js'
import { InferenceDoc } from '../../models/Inference.js'
import { ModelCardInterface, ModelDoc, ModelInterface } from '../../models/Model.js'
import { ReleaseDoc } from '../../models/Release.js'
import { ResponseInterface } from '../../models/Response.js'
import { ReviewInterface } from '../../models/Review.js'
import { SchemaDoc, SchemaInterface } from '../../models/Schema.js'
import { TokenDoc } from '../../models/Token.js'
import { ModelSearchResult } from '../../routes/v2/model/getModelsSearch.js'
import { BailoError } from '../../types/error.js'

const AuditKind = {
  Create: 'Create',
  View: 'View',
  Update: 'Update',
  Delete: 'Delete',
  Search: 'Search',
} as const
export type AuditKindKeys = (typeof AuditKind)[keyof typeof AuditKind]

export const AuditInfo = {
  CreateModel: { typeId: 'CreateModel', description: 'Model Created', auditKind: AuditKind.Create },
  ViewModel: { typeId: 'ViewModel', description: 'Model Viewed', auditKind: AuditKind.View },
  UpdateModel: { typeId: 'UpdateModel', description: 'Model Updated', auditKind: AuditKind.Update },
  SearchModels: { typeId: 'SearchModels', description: 'Model Searched', auditKind: AuditKind.Search },

  CreateModelCard: { typeId: 'CreateModelCard', description: 'Model Card Created', auditKind: AuditKind.Create },
  ViewModelCard: { typeId: 'ViewModelCard', description: 'Model Card Viewed', auditKind: AuditKind.View },
  ViewModelCardRevisions: {
    typeId: 'ViewModelCardRevisions',
    description: 'Model Card Revisions Viewed',
    auditKind: AuditKind.View,
  },
  UpdateModelCard: { typeId: 'UpdateModelCard', description: 'Model Card Updated', auditKind: AuditKind.Update },

  CreateFile: { typeId: 'CreateFile', description: 'File Information Created', auditKind: AuditKind.Create },
  ViewFile: { typeId: 'ViewFile', description: 'File Downloaded', auditKind: AuditKind.View },
  ViewFiles: { typeId: 'ViewFiles', description: 'File Information Viewed', auditKind: AuditKind.View },
  DeleteFile: { typeId: 'DeleteFile', description: 'File Information Deleted', auditKind: AuditKind.Delete },

  CreateRelease: { typeId: 'CreateRelease', description: 'Release Created', auditKind: AuditKind.Create },
  ViewRelease: { typeId: 'ViewRelease', description: 'Release Viewed', auditKind: AuditKind.View },
  UpdateRelease: { typeId: 'UpdateRelease', description: 'Release Updated', auditKind: AuditKind.Update },
  DeleteRelease: { typeId: 'DeleteRelease', description: 'Release Deleted', auditKind: AuditKind.Delete },
  ViewReleases: { typeId: 'ViewReleases', description: 'Releases Viewed', auditKind: AuditKind.View },

  CreateUserToken: { typeId: 'CreateUserToken', description: 'Token Created', auditKind: AuditKind.Create },
  ViewUserTokens: { typeId: 'ViewUserToken', description: 'Token Viewed', auditKind: AuditKind.View },
  DeleteUserToken: { typeId: 'DeleteUserToken', description: 'Token Deleted', auditKind: AuditKind.Delete },

  CreateAccessRequest: {
    typeId: 'CreateAccessRequest',
    description: 'Access Request Created',
    auditKind: AuditKind.Create,
  },
  ViewAccessRequest: { typeId: 'ViewAccessRequest', description: 'Access Request Viewed', auditKind: AuditKind.View },
  UpdateAccessRequest: {
    typeId: 'UpdateAccess Request',
    description: 'Access Request Updated',
    auditKind: AuditKind.Update,
  },
  DeleteAccessRequest: {
    typeId: 'UpdateAccessRequest',
    description: 'Access Request Deleted',
    auditKind: AuditKind.Update,
  },
  ViewAccessRequests: {
    typeId: 'ViewAccessRequests',
    description: 'Access Requests Viewed',
    auditKind: AuditKind.View,
  },

  SearchReviews: {
    typeId: 'SearchReviews',
    description: 'Reviews Searched',
    auditKind: AuditKind.Search,
  },
  CreateReviewResponse: {
    typeId: 'CreateReviewResponse',
    description: 'Review Response Created',
    auditKind: AuditKind.Create,
  },
  UpdateReviewResponse: {
    typeId: 'UpdateReviewResponse',
    description: 'Review Response Updated',
    auditKind: AuditKind.Update,
  },

  CreateSchema: { typeId: 'CreateSchema', description: 'Schema Created', auditKind: AuditKind.Create },
  SearchSchemas: { typeId: 'SearchedSchemas', description: 'Schemas Searched', auditKind: AuditKind.Search },
  ViewSchema: { typeId: 'ViewSchema', description: 'Schema Viewed', auditKind: AuditKind.View },
  DeleteSchema: { typeId: 'DeleteSchema', description: 'Schema Deleted', auditKind: AuditKind.Delete },
  UpdateSchema: { typeId: 'UpdateSchema', description: 'Schema Updated', auditKind: AuditKind.Update },

  ViewModelImages: { typeId: 'ViewModelImages', description: 'Model Images Viewed', auditKind: AuditKind.View },

  CreateInference: { typeId: 'CreateInference', description: 'Inference Service Created', auditKind: AuditKind.Create },
  UpdateInference: { typeId: 'UpdateInference', description: 'Inference Service Updated', auditKind: AuditKind.Update },
  ViewInference: { typeId: 'ViewInference', description: 'Inference Service Viewed', auditKind: AuditKind.View },
  ViewInferences: { typeId: 'ViewInferences', description: 'Inferences Viewed', auditKind: AuditKind.View },

  CreateExport: { typeId: 'CreateExport', description: 'Model Exported', auditKind: AuditKind.Create },
  CreateImport: { typeId: 'CreateImport', description: 'Model Imported', auditKind: AuditKind.Create },

  ViewResponses: {
    typeId: 'ViewResponses',
    description: 'View a list of responses',
    auditKind: AuditKind.View,
  },
  CreateResponse: {
    typeId: 'CreateResponse',
    description: 'Review or comment responses created',
    auditKind: AuditKind.Create,
  },
  UpdateResponse: {
    typeId: 'UpdateResponse',
    description: 'Updated a comment or review response',
    auditKind: AuditKind.Update,
  },
} as const
export type AuditInfoKeys = (typeof AuditInfo)[keyof typeof AuditInfo]

export abstract class BaseAuditConnector {
  abstract onCreateModel(req: Request, model: ModelDoc)
  abstract onViewModel(req: Request, model: ModelDoc)
  abstract onUpdateModel(req: Request, model: ModelDoc)
  abstract onSearchModel(req: Request, models: ModelSearchResult[])

  abstract onCreateModelCard(req: Request, modelId: string, modelCard: ModelCardInterface)
  abstract onViewModelCard(req: Request, modelId: string, modelCard: ModelCardInterface)
  abstract onUpdateModelCard(req: Request, modelId: string, modelCard: ModelCardInterface)
  abstract onViewModelCardRevisions(req: Request, modelId: string, modelCards: ModelCardInterface[])

  abstract onCreateFile(req: Request, file: FileInterfaceDoc)
  abstract onViewFile(req: Request, file: FileInterfaceDoc)
  abstract onViewFiles(req: Request, modelId: string, files: FileInterface[])
  abstract onDeleteFile(req: Request, modelId: string, fileId: string)

  abstract onCreateRelease(req: Request, release: ReleaseDoc)
  abstract onViewRelease(req: Request, release: ReleaseDoc)
  abstract onUpdateRelease(req: Request, release: ReleaseDoc)
  abstract onDeleteRelease(req: Request, modelId: string, semver: string)
  abstract onViewReleases(req: Request, releases: ReleaseDoc[])

  abstract onCreateCommentResponse(req: Request, response: ResponseInterface)
  abstract onCreateReviewResponse(req: Request, response: ResponseInterface)

  abstract onViewResponses(req: Request, responseInterfaces: ResponseInterface[])
  abstract onUpdateResponse(req: Request, responseId: string)

  abstract onCreateUserToken(req: Request, token: TokenDoc)
  abstract onViewUserTokens(req: Request, tokens: TokenDoc[])
  abstract onDeleteUserToken(req: Request, accessKey: string)

  abstract onCreateAccessRequest(req: Request, accessRequest: AccessRequestDoc)
  abstract onViewAccessRequest(req: Request, accessRequest: AccessRequestDoc)
  abstract onUpdateAccessRequest(req: Request, accessRequest: AccessRequestDoc)
  abstract onDeleteAccessRequest(req: Request, accessRequestId: string)
  abstract onViewAccessRequests(req: Request, accessRequests: AccessRequestDoc[])

  abstract onSearchReviews(req: Request, reviews: (ReviewInterface & { model: ModelInterface })[])

  abstract onSearchSchemas(req: Request, schemas: SchemaInterface[])
  abstract onCreateSchema(req: Request, schema: SchemaInterface)
  abstract onViewSchema(req: Request, schema: SchemaInterface)
  abstract onDeleteSchema(req: Request, schemaId: string)
  abstract onUpdateSchema(req: Request, schema: SchemaDoc)

  abstract onCreateInference(req: Request, inference: InferenceDoc)
  abstract onUpdateInference(req: Request, inference: InferenceDoc)
  abstract onViewInference(req: Request, inference: InferenceDoc)
  abstract onViewInferences(req: Request, inference: InferenceDoc[])

  abstract onViewModelImages(
    req: Request,
    modelId: string,
    images: { repository: string; name: string; tags: string[] }[],
  )

  abstract onCreateS3Export(req: Request, modelId: string, semvers?: string[])
  abstract onCreateImport(
    req: Request,
    mirroredModelId: string,
    sourceModelId: string,
    modelCardVersions: number[],
    exporter: string,
  )

  abstract onError(req: Request, error: BailoError)

  checkEventType(auditInfo: AuditInfoKeys, req: Request) {
    if (auditInfo.typeId !== req.audit.typeId && auditInfo.description !== req.audit.description) {
      throw new Error(`Audit: Expected type '${JSON.stringify(auditInfo)}' but recieved '${JSON.stringify(req.audit)}'`)
    }
  }
}
