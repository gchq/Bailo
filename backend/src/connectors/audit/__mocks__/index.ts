import { vi } from 'vitest'

const audit = {
  onCreateModel: vi.fn(),
  onViewModel: vi.fn(),
  onUpdateModel: vi.fn(),
  onSearchModel: vi.fn(),

  onCreateModelCard: vi.fn(),
  onViewModelCard: vi.fn(),
  onViewModelCardRevisions: vi.fn(),
  onUpdateModelCard: vi.fn(),

  onViewFiles: vi.fn(),
  onDeleteFile: vi.fn(),
  onCreateFile: vi.fn(),

  onCreateRelease: vi.fn(),
  onViewRelease: vi.fn(),
  onUpdateRelease: vi.fn(),
  onDeleteRelease: vi.fn(),
  onViewReleases: vi.fn(),

  onCreateAccessRequest: vi.fn(),
  onViewAccessRequest: vi.fn(),
  onUpdateAccessRequest: vi.fn(),
  onDeleteAccessRequest: vi.fn(),
  onViewAccessRequests: vi.fn(),

  onSearchReviews: vi.fn(),
  onCreateReviewResponse: vi.fn(),

  onCreateSchema: vi.fn(),
  onViewSchema: vi.fn(),
  onUpdateSchema: vi.fn(),
  onDeleteSchema: vi.fn(),
  onSearchSchemas: vi.fn(),

  onCreateInference: vi.fn(),
  onViewInference: vi.fn(),
  onUpdateInference: vi.fn(),
  onViewInferences: vi.fn(),

  onViewModelImages: vi.fn(),

  onCreateS3Export: vi.fn(),
  onCreateImport: vi.fn(),

  onCreateCommentResponse: vi.fn(),
  onViewResponses: vi.fn(),
  onUpdateResponse: vi.fn(),

  onError: vi.fn(),
}
export default audit
