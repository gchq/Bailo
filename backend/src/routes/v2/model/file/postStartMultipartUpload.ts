import bodyParser from 'body-parser'
import { Request, Response } from 'express'
import { z } from 'zod'

import { AuditInfo } from '../../../../connectors/audit/Base.js'
import audit from '../../../../connectors/audit/index.js'
import { startUploadMultipartFile } from '../../../../services/file.js'
import { registerPath } from '../../../../services/specification.js'
import { coerceArray, parse } from '../../../../utils/validate.js'

export const postStartMultipartUploadSchema = z.object({
  params: z.object({
    modelId: z.string(),
  }),
  body: z.object({
    name: z.string(),
    mime: z.string().optional().default('application/octet-stream'),
    size: z.number().positive(),
    tags: coerceArray(z.array(z.string()).optional()),
  }),
})

registerPath({
  method: 'post',
  path: '/api/v2/model/{modelId}/files/upload/multipart/start',
  tags: ['file'],
  description: 'Start uploading a new multipart file to a model.',
  schema: postStartMultipartUploadSchema,
  responses: {
    200: {
      description: "The newly created file's chunk details.",
      content: {
        'application/json': {
          schema: z.object({
            fileId: z.string().openapi({ example: '67cecbffd2a0951d1693b396' }),
            uploadId: z.string().openapi({ example: '67cecbffd2a0951d1693b396' }),
            chunks: z.array(z.object({ startByte: z.number(), endByte: z.number() })),
          }),
        },
      },
    },
  },
})

export interface ChunkByteRange {
  startByte: number
  endByte: number
}

interface PostStartMultipartUpload {
  fileId: string
  uploadId: string
  chunks: Array<ChunkByteRange>
}

export const postStartMultipartUpload = [
  bodyParser.json(),
  async (req: Request, res: Response<PostStartMultipartUpload>): Promise<void> => {
    req.audit = AuditInfo.CreateFile
    // Does user have permission to upload a file?
    const {
      params: { modelId },
      body: { name, mime, size, tags },
    } = parse(req, postStartMultipartUploadSchema)

    const { file, uploadId, chunks } = await startUploadMultipartFile(req.user, modelId, name, mime, size, tags)
    await audit.onCreateFile(req, file)

    res.json({
      fileId: file.id,
      uploadId,
      chunks,
    })
  },
]
