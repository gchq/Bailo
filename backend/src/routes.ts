import express from 'express'
import path from 'path'
import { fileURLToPath } from 'url'

import authentication from './connectors/authentication/index.js'
import { expressErrorHandler } from './routes/middleware/expressErrorHandler.js'
import { expressLogger } from './routes/middleware/expressLogger.js'
import { requestId } from './routes/middleware/requestId.js'
import { getDockerRegistryAuth } from './routes/v1/registryAuth.js'
import { getCurrentUser } from './routes/v2/entities/getCurrentUser.js'
import { getEntities } from './routes/v2/entities/getEntities.js'
import { getEntityLookup } from './routes/v2/entities/getEntityLookup.js'
import { getFilescanningInfo } from './routes/v2/filescanning/getFilescanningInfo.js'
import { putFileScan } from './routes/v2/filescanning/putFileScan.js'
import { deleteAccessRequest } from './routes/v2/model/accessRequest/deleteAccessRequest.js'
import { getAccessRequest } from './routes/v2/model/accessRequest/getAccessRequest.js'
import { getAccessRequestCurrentUserPermissions } from './routes/v2/model/accessRequest/getAccessRequestCurrentUserPermissions.js'
import { getModelAccessRequests } from './routes/v2/model/accessRequest/getModelAccessRequests.js'
import { patchAccessRequest } from './routes/v2/model/accessRequest/patchAccessRequest.js'
import { postAccessRequest } from './routes/v2/model/accessRequest/postAccessRequest.js'
import { postAccessRequestComment } from './routes/v2/model/accessRequest/postAccessRequestComment.js'
import { deleteFile } from './routes/v2/model/file/deleteFile.js'
import { getDownloadFile } from './routes/v2/model/file/getDownloadFile.js'
import { getFiles } from './routes/v2/model/file/getFiles.js'
import { postFinishMultipartUpload } from './routes/v2/model/file/postFinishMultipartUpload.js'
import { postSimpleUpload } from './routes/v2/model/file/postSimpleUpload.js'
import { postStartMultipartUpload } from './routes/v2/model/file/postStartMultipartUpload.js'
import { getModel } from './routes/v2/model/getModel.js'
import { getModelCurrentUserPermissions } from './routes/v2/model/getModelCurrentUserPermissions.js'
import { getModelsSearch } from './routes/v2/model/getModelsSearch.js'
import { getImages } from './routes/v2/model/images/getImages.js'
import { getInference } from './routes/v2/model/inferencing/getInferenceService.js'
import { getInferences } from './routes/v2/model/inferencing/getInferenceServices.js'
import { postInference } from './routes/v2/model/inferencing/postInferenceService.js'
import { putInference } from './routes/v2/model/inferencing/putInferenceService.js'
import { getModelCard } from './routes/v2/model/modelcard/getModelCard.js'
import { getModelCardHtml } from './routes/v2/model/modelcard/getModelCardHtml.js'
import { getModelCardRevisions } from './routes/v2/model/modelcard/getModelCardRevisions.js'
import { postFromSchema } from './routes/v2/model/modelcard/postFromSchema.js'
import { postFromTemplate } from './routes/v2/model/modelcard/postFromTemplate.js'
import { putModelCard } from './routes/v2/model/modelcard/putModelCard.js'
import { patchModel } from './routes/v2/model/patchModel.js'
import { postModel } from './routes/v2/model/postModel.js'
import { postRequestExportToS3 } from './routes/v2/model/postRequestExport.js'
import { postRequestImportFromS3 } from './routes/v2/model/postRequestImport.js'
import { getModelCurrentUserRoles } from './routes/v2/model/roles/getModelCurrentUserRoles.js'
import { getModelRoles } from './routes/v2/model/roles/getModelRoles.js'
import { deleteWebhook } from './routes/v2/model/webhook/deleteWebhook.js'
import { getWebhooks } from './routes/v2/model/webhook/getWebhooks.js'
import { postWebhook } from './routes/v2/model/webhook/postWebhook.js'
import { putWebhook } from './routes/v2/model/webhook/putWebhook.js'
import { deleteRelease } from './routes/v2/release/deleteRelease.js'
import { getRelease } from './routes/v2/release/getRelease.js'
import { getReleases } from './routes/v2/release/getReleases.js'
import { postRelease } from './routes/v2/release/postRelease.js'
import { postReleaseComment } from './routes/v2/release/postReleaseComment.js'
import { putRelease } from './routes/v2/release/putRelease.js'
import { getResponses } from './routes/v2/response/getResponses.js'
import { patchResponse } from './routes/v2/response/patchResponse.js'
import { patchResponseReaction } from './routes/v2/response/patchResponseReaction.js'
import { getReviews } from './routes/v2/review/getReviews.js'
import { postAccessRequestReviewResponse } from './routes/v2/review/postAccessRequestReviewResponse.js'
import { postReleaseReviewResponse } from './routes/v2/review/postReleaseReviewResponse.js'
import { deleteSchema } from './routes/v2/schema/deleteSchema.js'
import { getSchema } from './routes/v2/schema/getSchema.js'
import { getSchemas } from './routes/v2/schema/getSchemas.js'
import { patchSchema } from './routes/v2/schema/patchSchema.js'
import { postSchema } from './routes/v2/schema/postSchema.js'
import { getSpecification } from './routes/v2/specification.js'
import { getUiConfig } from './routes/v2/uiConfig/getUiConfig.js'
import { deleteUserToken } from './routes/v2/user/deleteUserToken.js'
import { getUserTokenList } from './routes/v2/user/getUserTokenList.js'
import { getUserTokens } from './routes/v2/user/getUserTokens.js'
import { postUserToken } from './routes/v2/user/postUserToken.js'
import config from './utils/config.js'

export const server = express()

server.use('/api/v2', requestId)
server.use('/api/v2', expressLogger)
const middlewareConfigs = authentication.authenticationMiddleware()
for (const middlewareConf of middlewareConfigs) {
  server.use(middlewareConf?.path || '/', middlewareConf.middleware)
}

server.get('/api/v1/registry_auth', ...getDockerRegistryAuth)

server.post('/api/v2/models', ...postModel)
server.get('/api/v2/models/search', ...getModelsSearch)
// server.post('/api/v2/models/import', ...postModelImport)

server.get('/api/v2/model/:modelId', ...getModel)
server.patch('/api/v2/model/:modelId', ...patchModel)

server.post('/api/v2/model/:modelId/export/s3', ...postRequestExportToS3)
server.post('/api/v2/model/import/s3', ...postRequestImportFromS3)

server.get('/api/v2/model/:modelId/model-card/:version', ...getModelCard)
server.get('/api/v2/model/:modelId/model-card/:version/html', ...getModelCardHtml)
server.get('/api/v2/model/:modelId/model-card-revisions', ...getModelCardRevisions)
server.put('/api/v2/model/:modelId/model-cards', ...putModelCard) // *server.get('/api/v2/template/models', ...getModelTemplates)
// *server.post('/api/v2/model/:modelId/setup/from-existing', ...postFromExisting)
server.post(`/api/v2/model/:modelId/setup/from-template`, ...postFromTemplate)
server.post('/api/v2/model/:modelId/setup/from-schema', ...postFromSchema)

server.post('/api/v2/model/:modelId/releases', ...postRelease)
server.get('/api/v2/model/:modelId/releases', ...getReleases)
server.get('/api/v2/model/:modelId/release/:semver', ...getRelease)
server.get('/api/v2/model/:modelId/release/:semver/file/:fileName/download', ...getDownloadFile)
// This is a temporary workaround to split out the URL to disable authorisation.
server.get('/api/v2/token/model/:modelId/release/:semver/file/:fileName/download', ...getDownloadFile)
server.put('/api/v2/model/:modelId/release/:semver', ...putRelease)
server.post('/api/v2/model/:modelId/release/:semver/comment', ...postReleaseComment)
server.delete('/api/v2/model/:modelId/release/:semver', ...deleteRelease)
server.post('/api/v2/model/:modelId/release/:semver/review', ...postReleaseReviewResponse)

server.post('/api/v2/model/:modelId/access-requests', ...postAccessRequest)
server.get('/api/v2/model/:modelId/access-requests', getModelAccessRequests)
server.get('/api/v2/model/:modelId/access-request/:accessRequestId', ...getAccessRequest)
server.delete('/api/v2/model/:modelId/access-request/:accessRequestId', ...deleteAccessRequest)
server.patch('/api/v2/model/:modelId/access-request/:accessRequestId', ...patchAccessRequest)
server.post('/api/v2/model/:modelId/access-request/:accessRequestId/comment', ...postAccessRequestComment)
server.post('/api/v2/model/:modelId/access-request/:accessRequestId/review', ...postAccessRequestReviewResponse)
server.get(
  '/api/v2/model/:modelId/access-request/:accessRequestId/permissions/mine',
  ...getAccessRequestCurrentUserPermissions,
)

server.get('/api/v2/model/:modelId/files', ...getFiles)
server.get('/api/v2/model/:modelId/file/:fileId/download', ...getDownloadFile)
// This is a temporary workaround to split out the URL to disable authorisation.
server.get('/api/v2/token/model/:modelId/file/:fileId/download', ...getDownloadFile)
server.post('/api/v2/model/:modelId/files/upload/simple', ...postSimpleUpload)
server.post('/api/v2/model/:modelId/files/upload/multipart/start', ...postStartMultipartUpload)
server.post('/api/v2/model/:modelId/files/upload/multipart/finish', ...postFinishMultipartUpload)
server.delete('/api/v2/model/:modelId/file/:fileId', ...deleteFile)

server.post('/api/v2/model/:modelId/webhooks', ...postWebhook)
server.get('/api/v2/model/:modelId/webhooks', ...getWebhooks)
server.put('/api/v2/model/:modelId/webhook/:webhookId', ...putWebhook)
server.delete('/api/v2/model/:modelId/webhook/:webhookId', ...deleteWebhook)

server.get('/api/v2/model/:modelId/images', ...getImages)
// *server.delete('/api/v2/model/:modelId/images/:imageId', ...deleteImage)
server.get('/api/v2/model/:modelId/files', ...getFiles)
server.get('/api/v2/model/:modelId/file/:fileId/download', ...getDownloadFile)
// This is a temporary workaround to split out the URL to disable authorisation.
server.get('/api/v2/token/model/:modelId/file/:fileId/download', ...getDownloadFile)
server.post('/api/v2/model/:modelId/files/upload/simple', ...postSimpleUpload)
server.post('/api/v2/model/:modelId/files/upload/multipart/start', ...postStartMultipartUpload)
server.post('/api/v2/model/:modelId/files/upload/multipart/finish', ...postFinishMultipartUpload)
server.delete('/api/v2/model/:modelId/file/:fileId', ...deleteFile)

// Inferencing routes are conditional to Bailo specific configuration
// NOTE: Enabled if config not loaded for testing purposes
if (!config.ui?.inference || config.ui.inference?.enabled) {
  server.get('/api/v2/model/:modelId/inferences', ...getInferences)
  server.get('/api/v2/model/:modelId/inference/:image/:tag', ...getInference)
  server.post('/api/v2/model/:modelId/inference', ...postInference)
  server.put('/api/v2/model/:modelId/inference/:image/:tag', ...putInference)
}
// *server.get('/api/v2/model/:modelId/release/:semver/file/:fileCode/list', ...getModelFileList)
// *server.get('/api/v2/model/:modelId/release/:semver/file/:fileCode/raw', ...getModelFileRaw)

server.get('/api/v2/schemas', ...getSchemas)
server.get('/api/v2/schema/:schemaId', ...getSchema)
server.post('/api/v2/schemas', ...postSchema)
server.patch('/api/v2/schema/:schemaId', ...patchSchema)
server.delete('/api/v2/schema/:schemaId', ...deleteSchema)

server.get('/api/v2/reviews', ...getReviews)
server.get('/api/v2/responses', ...getResponses)
server.patch('/api/v2/response/:responseId', ...patchResponse)
server.patch('/api/v2/response/:responseId/reaction/:kind', ...patchResponseReaction)

server.get('/api/v2/model/:modelId/roles', ...getModelRoles)
server.get('/api/v2/model/:modelId/roles/mine', ...getModelCurrentUserRoles)
server.get('/api/v2/model/:modelId/permissions/mine', ...getModelCurrentUserPermissions)

server.get('/api/v2/entities', ...getEntities)
server.get('/api/v2/entities/me', ...getCurrentUser)
server.get('/api/v2/entity/:dn/lookup', ...getEntityLookup)

server.get('/api/v2/config/ui', ...getUiConfig)

server.post('/api/v2/user/tokens', ...postUserToken)
server.get('/api/v2/user/tokens', ...getUserTokens)
server.get('/api/v2/user/tokens/list', ...getUserTokenList)
// server.get('/api/v2/user/:userId/token/:tokenId', ...getUserToken)
server.delete('/api/v2/user/token/:accessKey', ...deleteUserToken)

server.get('/api/v2/specification', ...getSpecification)

server.get('/api/v2/filescanning/info', ...getFilescanningInfo)
server.put('/api/v2/filescanning/model/:modelId/file/:fileId/scan', ...putFileScan)

// Python docs
const __filename = fileURLToPath(import.meta.url)
const __dirname = path.dirname(__filename)
server.use('/docs/python', express.static(path.join(__dirname, '../python-docs/dirhtml')))

server.use('/api/v2', expressErrorHandler)
