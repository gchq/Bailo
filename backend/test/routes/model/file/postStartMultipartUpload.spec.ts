import { describe, expect, test, vi } from 'vitest'

import audit from '../../../../src/connectors/audit/__mocks__/index.js'
import { postStartMultipartUploadSchema } from '../../../../src/routes/v2/model/file/postStartMultipartUpload.js'
import { createFixture, testPost } from '../../../testUtils/routes.js'

vi.mock('../../../../src/utils/user.js')
vi.mock('../../../../src/connectors/audit/index.js')
vi.mock('../../../../src/connectors/authorisation/index.js')

vi.mock('../../../../src/services/file.js', () => ({
  startUploadMultipartFile: vi.fn(() => ({
    file: { id: 'test' },
    uploadId: 'test',
    chunks: [{ presignedUrl: 'string', startByte: 0, endByte: 12345 }],
  })),
}))

describe('routes > model > file > postStartMultipartUpload', () => {
  test('200 > ok', async () => {
    const fixture = createFixture(postStartMultipartUploadSchema)
    const res = await testPost(`/api/v2/model/${fixture.params.modelId}/files/upload/multipart/start`, fixture)

    expect(res.statusCode).toBe(200)
    expect(res.body).matchSnapshot()
  })

  test('audit > expected call', async () => {
    const fixture = createFixture(postStartMultipartUploadSchema)
    const res = await testPost(`/api/v2/model/${fixture.params.modelId}/files/upload/multipart/start`, fixture)

    expect(res.statusCode).toBe(200)
    expect(audit.onCreateFile).toBeCalled()
    expect(audit.onCreateFile.mock.calls.at(0)?.at(1)).toMatchSnapshot()
  })

  test('400 > no name', async () => {
    const fixture = createFixture(postStartMultipartUploadSchema) as any
    // This fixture does not include a name.
    delete fixture.body.name

    const res = await testPost(`/api/v2/model/${fixture.params.modelId}/files/upload/multipart/start`, fixture)

    expect(res.statusCode).toEqual(400)
    expect(res.body).matchSnapshot()
  })
})
