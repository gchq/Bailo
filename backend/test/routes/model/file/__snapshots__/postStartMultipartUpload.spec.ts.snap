// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`routes > model > file > postStartMultipartUpload > 200 > ok 1`] = `
{
  "chunks": [
    {
      "endByte": 12345,
      "presignedUrl": "string",
      "startByte": 0,
    },
  ],
  "fileId": "test",
  "uploadId": "test",
}
`;

exports[`routes > model > file > postStartMultipartUpload > 400 > no name 1`] = `
{
  "error": {
    "context": {
      "errors": [
        {
          "code": "invalid_type",
          "expected": "string",
          "message": "Required",
          "path": [
            "body",
            "name",
          ],
          "received": "undefined",
        },
      ],
    },
    "message": "Path: body.name - Message: Required",
    "name": "Bailo Error",
  },
}
`;

exports[`routes > model > file > postStartMultipartUpload > audit > expected call 1`] = `
{
  "id": "test",
}
`;
